plaso/parsers/megasync.py

```diff
@@ -0,0 +1,203 @@
+      ('line', _LOG_LINE),
+      ('repeat', _REPEAT_LINE),
+      ('program_start', _PROGRAM_START)]
+
+  _LINES_OF_INTEREST = [
+    'Transfer (UPLOAD) finished',
+    'Transfer (UPLOAD) starting',
+    'Upload complete',
+    'Creating thumb/preview'
+  ]
+
+  def __init__(self):
+    """Initializes a parser."""
+    super(MEGASyncParser, self).__init__()
+    self._last_month = 0
+    self._maximum_year = 0
+    self._year_use = 0
+
+  def _UpdateYear(self, mediator, month):
+    """Updates the year to use for events, based on last observed month.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      month (int): month observed by the parser, where January is 1.
+    """
+    # TODO: Investigate using timestamps of the Gzip file, not the file
+    # within the Gzip file as the basis of estimation.
+    if not self._year_use:
+      self._year_use = mediator.GetEstimatedYear()
+
+      # zlib (used by MEGASync to compress rotated-out log files)
+      # can generate Gzip files with empty modification timestamp.
+      # This shouldn't be used as the estimated year.
+      # MEGASync logs can't originate from 1970, so this is safe.
+      if self._year_use == 1970:
+        self._year_use = mediator.GetCurrentYear()
+    if not self._maximum_year:
+      self._maximum_year = mediator.GetLatestYear()
+
+    if not self._last_month:
+      self._last_month = month
+      return
+
+    if self._last_month > month:
+      if self._year_use < self._maximum_year:
+        self._year_use += 1
+    self._last_month = month
+
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a structure of tokens derived from a line of a text file.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): structure of tokens derived from
+          a line of a text file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key == 'line':
+      time_elements_tuple = self._GetValueFromStructure(structure, 'timestamp')
+      month, day_of_month, hours, minutes, seconds, microseconds = (
+          time_elements_tuple)
+
+      self._UpdateYear(parser_mediator, month)
+
+      time_elements_tuple = (
+          self._year_use,
+          month, day_of_month, hours, minutes, seconds, microseconds)
+
+      try:
+        timestamp = dfdatetime_time_elements.TimeElementsInMicroseconds(
+            time_elements_tuple=time_elements_tuple)
+      except ValueError:
+        parser_mediator.ProduceExtractionWarning(
+            'invalid timestamp: {0!s}'.format(time_elements_tuple)
+        )
+        return
+
+      log_message = self._GetValueFromStructure(structure, 'message')
+
+      for line in self._LINES_OF_INTEREST:
```

Comment:
MEGASync logs are very verbose, so I currently only produce events for log lines that may be of forensic interest. If it's more desirable to produce events for every log entry, I can change this, but I recommend checking out the sample test file for context. (also [here](https://gist.github.com/hur/f185cee1976b4773a1c8defa2fedef43) if the diff viewer wont show it)

Note that in an environment where MEGAsync ran for more than a few minutes, you will likely have 50 of those log files, totaling some ~4 million lines.